### ISSUE TITLE
Heartbeat for Notifications causes too many CPU executions

### ISSUE THREAD

**Issue by AndreaBarghigiani** [gocodebox/lifterlms#371]
I've noticed that when I am logged inside my WordPress installation inside the footer is been injected the following code:
```
<script id="llms-inline-footerscripts" type="text/javascript">
    window.llms = window.llms || {};
    window.llms.notification_settings = {"heartbeat_interval":20000};
    window.llms = window.llms || {};
    window.llms.ajaxurl = "http://mysite.url/wp-admin/admin-ajax.php";
    window.LLMS = window.LLMS || {};
    window.LLMS.l10n = window.LLMS.l10n || {};
    window.LLMS.l10n.strings = {...};
</script>
```

I would like to understand why this is happening because more logged in user I have more are the CPU execution for `http://mysite.url/wp-admin/admin-ajax.php` (I've masked the domain name) and my hosting does not allow me to run so many requests.

I would like to understand if I am able to reduce this kind of request and if so how or at least how to disable on a specific page of my site. For example, I would like to run them only on the lesson pages.

If is possible I would also like to understand how I can disable some notifications on the frontend, the Base and not the Email, because I do not want to use them. I see on the backend that in each configuration I am able to subscribe the kind of user as the following image: 

![schermata 2017-08-25 alle 18 13 03](https://user-images.githubusercontent.com/190159/29722442-1a6e76d6-89c1-11e7-993d-e6400298c050.png)

If I disable the checkbox will I be able to reduce the `/wp-admin/admin-ajax.php` calls?

Thank you in advance.
All the best Andrea

PS: in the meantime I have used the `llms_notifications_settings` filter and I've set the `$settings['heartbeat_interval']` variable to `300000` that I believe is 300 seconds = 5 minutes. Is it right?

**Comment by thomasplevy** [comment]
@AndreaBarghigiani I am aware of the CPU issues here. During beta and testing not a single person raised this as a potential issue and now that it's been released I have been receiving complaints from users on shared hosts, mostly. Technically we're not doing anything incorrectly here, though it is a performance concern for larger sites and I wish that would have occurred to me prior to release. The way WordPress __recommends__ making AJAX calls is via admin-ajax.php and unfortunately admin-ajax.php boots the whole admin panel without any caching and is not performant.

I have been spending a lot of mental energy attempting to figure out a better way to make AJAX requests from the frontend of a WordPress website. The REST API is the answer but it requires a lot of redoing of the codebase (we don't have any REST API functionality built into LifterLMS currently). This is high on my radar but as a lone developer I have a lot of things high on my radar and while this is important it's not something I've prioritized development of yet.

Using the filter to cut down on the frequency of your requests is the only interim solution I have for you.

You may find a performance inscrease if you tweak the autoloaded data on your site. I find that a lot of sites I've touched over the years autoload a ton too much data and some plugins very liberally set options to be autoloaded that really don't need to be. This data means a lot of db calls on ajax calls and that could mean high CPU usage. 

For example, on LifterLMS.com we use Mandrill to send our transactional emails and we use a plugin to handle that. I found that the plugin was saving a bunch of statistical data (which I frankly don't care a lick about) as a giant serialized array as an autoloaded option. Deleting that option and then turning off the stats options in the plugin had a pretty great performance improvement for my site.

I also went through the database and found I had smaller autoloaded options from plugins we no longer used (some of which had been installed and deleted over 4 years ago).

This query  will show you the 20 largest (in bytes) autoloaded options on your site:

```sql
SELECT LENGTH(option_value),option_name FROM wp_options WHERE autoload='yes' ORDER BY length(option_value) DESC LIMIT 20;
```

And this one will give you a total (in bytes) of all of the autoloaded options on your site:

```sql
-- less than 800k
SELECT SUM(LENGTH(option_value)) FROM wp_options WHERE autoload = 'yes';
```

A very smart sysadmin gave me this and noted he always shoots to get this lower than 800k. I find it's not that hard to get sites under 400k.

I know this won't solve the LifterLMS core problems but I hope it can help in the meantime while I figure out a better way to handle this in a future release

**Comment by Mte90** [comment]
Probably if it is only for notification is better to use the rest api than the heartbeat that is used for other stuff (auto save etc).
Looking on the net the issue was already discussed:

* https://deliciousbrains.com/comparing-wordpress-rest-api-performance-admin-ajax-php/
* https://woorkup.com/diagnose-admin-ajax-php-causing-slow-load-times-wordpress/

**Comment by EugenioPetulla** [comment]
Hi @thomasplevy I'm a @AndreaBarghigiani and @Mte90 co-worker and actually I'm working on some debugging about our LMS based project.

I discovered the problem today by analyzing the 500 errors we get in the past. Your solution based on filtering the setInterval time is good enough to me for a temporary fix (we will set it to 300000) and this should mitigate our issue.

I want spent more time in order to figure out how to integrate a notification system that doesn't rely on `admin-ajax.php` and REST API neither so maybe I can contribute to LMS in a more active way.

I want just ask you if is possible to have the assets folder unminified here on github because it caused me a big headache reading and manipulating beautified code that comes from your minifier 😃 

This could potentially help you also to gather more contributors around the project. You can just drop the script into the folder without registering it in WordPress or make a separate branch just for JS files that you automatically minify, I think, when you make the build and push on GitHub.

Leave this thread open please, and I will come here with some proposal in the near future. 😸

**Comment by thomasplevy** [comment]
@Mte90 you are 200% correct and hindsight is 20/20. It very much seemed like the best solution during development and I very much want to develop a LifterLMS REST API but I haven't prioritized it yet.

**Comment by thomasplevy** [comment]
@EugenioPetulla the unminified scripts can be found here: https://github.com/gocodebox/lifterlms/tree/master/_private/js

The build process around JS and SCSS is a mess that I simply haven't had the stomach to refactor so while I'd love to say "YES I CAN GIVE YOU UNMINIFIED SCRIPTS" I simply can't turn that around over night. It's a long road to rebuild the asset pipe without also breaking it.

In the meantime insall this for development and modify the unminified stuff in the assets directory. See here: https://github.com/gocodebox/lifterlms#installing-for-development

I am desperate for contributors as LifterLMS grows so any help you can provide would be very helpful.

The code in question is in this file specifically: https://github.com/gocodebox/lifterlms/blob/master/_private/js/llms-notifications.js

**Comment by EugenioPetulla** [comment]
Don't worry, I know that sensation so I can only say to be patient because we want contribute more. Our project is based on your plugin and we are 3 WordPress developers that are tryng to launch an e-learning for developers! 😄 

You already done a great job and we have just to fix something. I'm going to install your developer version so count on us for contribution! 👍

**Comment by thomasplevy** [comment]
@EugenioPetulla love it thank you!

**Comment by EugenioPetulla** [comment]
According to my investigation I think I can start to work on a REST API layer for your plugin and after I we can discuss on how to switch from AJAX notification to REST API.

Feel free to close this issue. We'll keep on touch later when I have some code to show to you. 😈

**Comment by thomasplevy** [comment]
@EugenioPetulla I'll leave this open for the time being until we have either a PR from you or us together to resolve the issue. It's something that's been nagging me (somewhat privately) since the 3.8 release and it's good to have the issue out in the public for anyone else to find, comment, or contribute to/about.

Keep me posted if you have any needs from me, especially with regards to the JS stuff. If that's causing you tremendous headache I can try to get something quickly patched in to make your development life easier

**Comment by EugenioPetulla** [comment]
I was trying to understand if you used the internal WordPress heartbeat script or if you managed to implement it by yourself so I beautified the minified JS and I found the setInterval function.

I made this because the internal heartbeat script provided by WordPress has a limit and it's impossibile to set the frequency rate higher than 120 seconds. Implementing everything by yourself you overcame the limit and it's vital to us because we need a very more high frequency rate in order to avoid exceeding the server's resources limit.

:)

**Comment by thomasplevy** [comment]
@EugenioPetulla sorry if that wasn't clear. Yes, I wrote a custom heartbeat (we're going to be using it for other things in the future). I considered using the internal heartbeat but I found it too limiting and from the start I knew it would all eventually be moved into a rest api of some kind so I wanted to ensure it had as little baggage as possible.

Are you not able to use the unminified JS for reviewing? I don't want you struggling to read uglified code. The JS is unminified in this directory: https://github.com/gocodebox/lifterlms/tree/master/_private/js

You can view all the notifications related code here: https://github.com/gocodebox/lifterlms/blob/master/_private/js/llms-notifications.js (This is the heartbeat class & functionality that I wrote, unminified!)

**Comment by thomasplevy** [comment]
ideally, if a rest api exists, we should be able to modify this section to make an ajax request to the rest api endpoint as opposed to the ajax handler: https://github.com/gocodebox/lifterlms/blob/master/_private/js/llms-notifications.js#L83-L123

This is the handler (powered by admin-ajax.php): https://github.com/gocodebox/lifterlms/blob/master/includes/class.llms.ajax.handler.php#L163-L189

**Comment by thomasplevy** [comment]
@EugenioPetulla curious if you've made any progress here? I'm getting ready to start working on speccing and building out a full REST API in the next few weeks and if you've made any progress would love to work on collaborating rather than starting from scratch

**Comment by Mte90** [comment]
Actually we never had the time to work on that :-/

**Comment by thomasplevy** [comment]
I'll be getting to work on this, hopefully, in the next few weeks then. If ya'll jump into it before then please let me know so we don't duplicate efforts. Otherwise I'll assume it's on us internally!

Best

**Comment by thomasplevy** [comment]
@AndreaBarghigiani, @Mte90, @EugenioPetulla 

Anyone interested in testing / benchmarking this initial rewrite which switches from admin-ajax.php calls to REST API ajax calls for the notification heartbeat?

"Stable" alpha here: https://github.com/gocodebox/lifterlms/releases/tag/3.15.0-alpha.1

I'd love another set of eyes or two on this initial build. I haven't implemented a fully-fledged REST API but have the basis in place to switch this set of calls over.

My benchmarking and testing (locally) shows this to be overall less CPU heavy than calls to admin-ajax.php but it doesn't seem to as drastic a jump in performance as I'd like it to be.

None the less, it's a step in the correct direction and I'd love some feedback and more eyeballs on this if any of you still care about this issue.

Thanks,

**Comment by Mte90** [comment]
Actually we have no time to try it :-/

**Comment by thomasplevy** [comment]
@Mte90 I know this has been hanging for a long time and I'm really sorry about that... 

@actual-saurabh came up with a pretty good solution to entirely disable popover (basic) notifications should you want to give it a shot: https://gist.github.com/actual-saurabh/9826484ca095e3f3c0430ca59ad853b3

**Comment by Mte90** [comment]
We was disabling compltelly heartbeat, this code is better :-)

**Comment by thomasplevy** [comment]
@Mte90 & @AndreaBarghigiani can either of you take a look [issue/371](https://github.com/gocodebox/lifterlms/tree/issue/371) branch which is a stab at removing AJAX-powered notifications entirely. All queries occur in PHP which removes real-time async notifications but improves performance by significantly reducing the number of requests made to your server.

I'm going to continue testing on this and if it satisfies your initial report and I get positive feedback from the team and some other affected users I'll merge it into core in a future release.

Thanks

**Comment by thomasplevy** [comment]
@Mte90 the UX for a notification should remain the same with the exception that users will only ever see notifications on page load -- there will be no more ajax polling for new notifications while a student is on page.

This adds additional database queries on page load (and yes, I know you hate that) but it reduces ajax calls.

Overall this will reduce strain on busy servers by significantly reducing the overall number of requests made.

**Comment by Mte90** [comment]
We are testing that on our production site, in case of issues I will revert :-)
